lib/mime.c

```diff
@@ -279,29 +280,66 @@ static void mimesetstate(struct mime_state *state,
 
 
 /* Escape header string into allocated memory. */
-static char *escape_string(const char *src)
+static CURLcode escape_mime_string(struct dynbuf *db, const char *src)
 {
-  size_t bytecount = 0;
-  size_t i;
-  char *dst;
+  CURLcode result = CURLE_OK;
 
-  for(i = 0; src[i]; i++)
-    if(src[i] == '"' || src[i] == '\\')
-      bytecount++;
+  for(; !result && *src; result = Curl_dyn_addn(db, src++, 1))
+    if(*src == '\\' || *src == '"') {
+      result = Curl_dyn_add(db, "\\");
+      if(result)
+        break;
+    }
+  return result;
+}
 
-  bytecount += i;
-  dst = malloc(bytecount + 1);
-  if(!dst)
-    return NULL;
+static CURLcode escape_form_string(struct Curl_easy *easy,
+                                   struct dynbuf *db, const char *src)
+{
+  CURLcode result = CURLE_OK;
+
+  (void) easy;
 
-  for(i = 0; *src; src++) {
-    if(*src == '"' || *src == '\\')
-      dst[i++] = '\\';
-    dst[i++] = *src;
+  for(; !result && *src; src++) {
+    if(*src != '"' && *src != '\r' && *src != '\n')
+      result = Curl_dyn_addn(db, src, 1);
+    else {
+      char ascchar = *src;
+
+#ifdef CURL_DOES_CONVERSIONS
+      result = Curl_convert_to_network(easy, &ascchar, 1);
+      if(result)
+        break;
+#endif
+      result = Curl_dyn_addf(db, "%%%02X", (unsigned char) ascchar);
+
+    }
   }
+  return result;
+}
+
+static char *escape_string(struct Curl_easy *easy,
+                           const char *src, enum mimestrategy strategy)
+{
+  CURLcode result;
+  struct dynbuf db;
 
-  dst[i] = '\0';
-  return dst;
+  Curl_dyn_init(&db, CURL_MAX_INPUT_LENGTH);
+  result = Curl_dyn_add(&db, "");
+
+  if(!result) {
+    if(strategy == MIMESTRATEGY_MAIL ||
+       (easy && easy->set.form_escape_as_mime))
```

Comment:
Can this function actually legitimately get called with `easy == NULL` ? If so, I think it requires a comment describing how/why. (and it should probably be named `data` for consistency).

Reply:
Yes, from `curl_formget` :-(

Reply:
Commented about `curl_formget`.